experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp

```diff
@@ -1158,6 +1158,92 @@ static Value buildStateCreation(TabularViewToStreamOp op,
                                             input);
 }
 
+//===----------------------------------------------------------------------===//
+// ValueToStreamOp.
+//===----------------------------------------------------------------------===//
+
+/// Builds IR that sets `hasReturned` to false. Possible output:
+///
+/// %3 = iterators.insertvalue %false into %arg0[1] : !iterators.state<i1, i32>
+static Value buildOpenBody(ValueToStreamOp op, OpBuilder &builder,
+                           Value initialState,
+                           ArrayRef<IteratorInfo> /*upstreamInfos*/) {
+  Location loc = op.getLoc();
+  ImplicitLocOpBuilder b(loc, builder);
+
+  // Reset hasReturned to false.
+  Value constFalse = b.create<arith::ConstantIntOp>(/*value=*/0, /*width=*/1);
+  Value updatedState = b.create<iterators::InsertValueOp>(
+      initialState, b.getIndexAttr(0), constFalse);
+
+  return updatedState;
+}
+
+/// Builds IR that returns the value in the first call and end-of-stream
+/// otherwise. Pseudo-code:
+///
+/// if hasReturned: return {}
+/// return value
+///
+/// Possible output:
+///
+/// %0 = iterators.extractvalue %arg0[0] : !iterators.state<i1, i32>
+/// %true = arith.constant true
+/// %1 = arith.xori %true, %0 : i1
+/// %2 = iterators.extractvalue %arg0[1] : !iterators.state<i1, i32>
+/// %3 = iterators.insertvalue %true into %arg0[0] : !iterators.state<i1, i32>
+static llvm::SmallVector<Value, 4>
+buildNextBody(ValueToStreamOp op, OpBuilder &builder, Value initialState,
+              ArrayRef<IteratorInfo> upstreamInfos, Type elementType) {
+  Location loc = op.getLoc();
+  ImplicitLocOpBuilder b(loc, builder);
+  Type i1 = b.getI1Type();
+
+  // Check if the iterator has returned an element already (since it should
+  // return one only in the first call to next).
+  Value hasReturned =
```

Comment:
this looks redundant with other things, can this be extracted to a common place and take a lambda somewhere ?